crates/ruff_db/src/vendored.rs

```diff
@@ -0,0 +1,351 @@
+use std::cell::RefCell;
+use std::io::{self, Read};
+use std::sync::{Mutex, MutexGuard};
+
+use itertools::Itertools;
+use zip::{read::ZipFile, ZipArchive};
+
+use crate::metadata::FileRevision;
+pub use path::{VendoredPath, VendoredPathBuf};
+
+pub mod path;
+
+type Result<T> = io::Result<T>;
+
+#[derive(Debug)]
+pub struct VendoredFileSystem {
+    inner: VendoredFileSystemInner,
+}
+
+impl VendoredFileSystem {
+    pub fn new(raw_bytes: &'static [u8]) -> Result<Self> {
+        Ok(Self {
+            inner: VendoredFileSystemInner::new(raw_bytes)?,
+        })
+    }
+
+    pub fn exists(&self, path: &VendoredPath) -> bool {
+        let normalized = normalize_vendored_path(path);
+        let inner_locked = self.inner.lock();
+        let mut archive = inner_locked.borrow_mut();
+        archive.lookup_path(&normalized).is_ok()
+            || archive
+                .lookup_path(&normalized.with_trailing_slash())
+                .is_ok()
```

Comment:
I would find some documentation helpful that explains the difference between paths ending with trailing slash and once that don't. Because it isn't clear from reading this code what's happening here and why proving the zip two times is necesary.